docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,721 @@
+
+/**
+ * The S3 Transfer Manager is a library that allows users to easily and
+ * optimally upload and downloads to and from S3.
+ * <p>
+ * The list of features includes:
+ * <ul>
+ *   <li>Parallel uploads and downloads</li>
+ *   <li>Bandwidth limiting</li>
+ *   <li>Pause and resume of transfers</li>
+ * </ul>
+ * <p>
+ * <b>Usage Example:</b>
+ * <pre>
+ * {@code
+ * // Create using all default configuration values
+ * S3TransferManager tm = S3TranferManager.create();
+ *
+ * // Using custom configuration values to set max upload speed to avoid
+ * // saturating the network interface.
+ * S3TransferManager tm = S3TransferManager.builder()
+ *         .configuration(TransferManagerConfiguration.builder()
+ *             .maximumUploadBytesSecond(32 * 1024 * 1024) // 32 MiB
+ *             .build()
+ *         .build();
+ * }
+ * </pre>
+ */
+public interface S3TransferManager {
+    /**
+     * DownloadObject an object identified by the bucket and key from S3 to the
+     * given file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * DownloadObject myFileDownload = tm.downloadObject(BUCKET, KEY, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload().completionFuture().join();
+     * }
+     * </pre>
+     *
+     */
+    default DownloadObject downloadObject(String bucket, String key, Path file) {
+        return downloadObject(DownloadObjectRequest.builder()
+                        .getObjectRequest(GetObjectRequest.builder()
+                                .bucket(bucket)
+                                .key(key)
+                                .build())
+                        .build(),
+                file);
+    }
+
+    /**
+     * DownloadObject an object using an S3 presigned URL to the given file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * DownloadObject myFileDownload = tm.downloadObject(myPresignedUrl, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload()completionFuture().join();
+     * }
+     * </pre>
+     */
+    default DownloadObject downloadObject(URL presignedUrl, Path file) {
+        return downloadObject(DownloadObjectRequest.builder()
+                        .presignedUrl(presignedUrl)
+                        .build(),
+                file);
+    }
+    /**
+     * DownloadObject an object in S3 to the given file.
+     *
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate the transfer
+     * DownloadObject myDownload = tm.downloadObject(DownloadObjectRequest.builder()
+     *         .getObjectRequest(GetObjectRequest.builder()
+     *                 .bucket(BUCKET)
+     *                 .key(KEY)
+     *                 .build())
+     *          // Set the known length of the object to avoid a HeadObject call
+     *         .hints(DownloadHints.builder()
```

Comment:
Hint sounds like it might not be accurate to me...

Reply:
Yeah it's a little funky but it is used in some places like SQL queries https://en.wikipedia.org/wiki/Hint_%28SQL%29.

Maybe "KnownProperties" or something along those lines?